File: 3762.minimum-operations-to-equalize-subarrays.py
#
# @lc app=leetcode id=3762 lang=python3
#
# [3762] Minimum Operations to Equalize Subarrays
#

# @lc code=start
from typing import List
import bisect

class Solution:
    def minOperations(self, nums: List[int], k: int, queries: List[List[int]]) -> List[int]:
        n = len(nums)
        mods = [num % k for num in nums]
        M = [num // k for num in nums]
        prefix_M = [0] * (n + 1)
        for i in range(n):
            prefix_M[i + 1] = prefix_M[i] + M[i]

        # Sparse table for min/max on mods
        LOG = 17
        st_min = [[0] * n for _ in range(LOG)]
        st_max = [[0] * n for _ in range(LOG)]
        for i in range(n):
            st_min[0][i] = mods[i]
            st_max[0][i] = mods[i]
        for j in range(1, LOG):
            pw = 1 << j
            for i in range(n - pw + 1):
                st_min[j][i] = min(st_min[j - 1][i], st_min[j - 1][i + (pw >> 1)])
                st_max[j][i] = max(st_max[j - 1][i], st_max[j - 1][i + (pw >> 1)])

        # logg for sparse queries
        logg = [0] * (n + 2)
        for i in range(2, n + 1):
            logg[i] = logg[i // 2] + 1

        # Merge sort tree
        tree = [[] for _ in range(4 * n)]
        prefixs = [[] for _ in range(4 * n)]

        def build(node: int, start: int, end: int) -> None:
            if start == end:
                tree[node] = [M[start]]
                prefixs[node] = [0, M[start]]
                return
            mid = (start + end) // 2
            build(2 * node, start, mid)
            build(2 * node + 1, mid + 1, end)
            left = tree[2 * node]
            right = tree[2 * node + 1]
            merged_len = len(left) + len(right)
            merged = [0] * merged_len
            i = j = kk = 0
            while i < len(left) and j < len(right):
                if left[i] <= right[j]:
                    merged[kk] = left[i]
                    i += 1
                else:
                    merged[kk] = right[j]
                    j += 1
                kk += 1
            while i < len(left):
                merged[kk] = left[i]
                i += 1
                kk += 1
            while j < len(right):
                merged[kk] = right[j]
                j += 1
                kk += 1
            tree[node] = merged
            pref = [0] * (merged_len + 1)
            for ii in range(merged_len):
                pref[ii + 1] = pref[ii] + merged[ii]
            prefixs[node] = pref

        def _query(node: int, start: int, end: int, ql: int, qr: int, val: int) -> tuple[int, int]:
            if ql > end or qr < start:
                return 0, 0
            if ql <= start and end <= qr:
                lst = tree[node]
                p = bisect.bisect_right(lst, val)
                return p, prefixs[node][p]
            mid = (start + end) // 2
            lc, ls = _query(2 * node, start, mid, ql, qr, val)
            rc, rs = _query(2 * node + 1, mid + 1, end, ql, qr, val)
            return lc + rc, ls + rs

        def query_leq(ql: int, qr: int, val: int) -> tuple[int, int]:
            return _query(1, 0, n - 1, ql, qr, val)

        def find_med(ql: int, qr: int) -> int:
            sz = qr - ql + 1
            target = sz // 2 + 1
            low = 0
            high = 10**9 + 10
            while low < high:
                midv = low + (high - low) // 2
                cnt, _ = query_leq(ql, qr, midv)
                if cnt >= target:
                    high = midv
                else:
                    low = midv + 1
            return low

        build(1, 0, n - 1)

        ans = []
        for li, ri in queries:
            leng = ri - li + 1
            j = logg[leng]
            mn = min(st_min[j][li], st_min[j][ri - (1 << j) + 1])
            mx = max(st_max[j][li], st_max[j][ri - (1 << j) + 1])
            if mn != mx:
                ans.append(-1)
                continue
            total_s = prefix_M[ri + 1] - prefix_M[li]
            sz = leng
            med = find_med(li, ri)
            cnt_lt, sum_lt = query_leq(li, ri, med - 1)
            cnt_leq, sum_leq = query_leq(li, ri, med)
            cnt_gt = sz - cnt_leq
            sum_gt = total_s - sum_leq
            cost = cnt_lt * med - sum_lt + sum_gt - cnt_gt * med
            ans.append(cost)
        return ans

# @lc code=end

File: 3762.minimum-operations-to-equalize-subarrays.cpp
#
# @lc app=leetcode id=3762 lang=cpp
#
# [3762] Minimum Operations to Equalize Subarrays
#
# @lc code=start
class Solution {
public:
    vector<long long> minOperations(vector<int>& nums, int k, vector<vector<int>>& queries) {
        int n = nums.size();
        int q = queries.size();
        
        vector<int> rem(n), quo(n);
        for (int i = 0; i < n; i++) {
            rem[i] = nums[i] % k;
            quo[i] = nums[i] / k;
        }
        
        vector<int> sortedQuo = quo;
        sort(sortedQuo.begin(), sortedQuo.end());
        sortedQuo.erase(unique(sortedQuo.begin(), sortedQuo.end()), sortedQuo.end());
        int m = sortedQuo.size();
        
        auto getIdx = [&](int v) -> int {
            return lower_bound(sortedQuo.begin(), sortedQuo.end(), v) - sortedQuo.begin() + 1;
        };
        
        vector<long long> countBIT(m + 2, 0), sumBIT(m + 2, 0);
        
        auto updateBIT = [&](vector<long long>& bit, int i, long long delta) {
            for (; i <= m + 1; i += i & (-i))
                bit[i] += delta;
        };
        
        auto queryBIT = [&](vector<long long>& bit, int i) -> long long {
            long long result = 0;
            for (; i > 0; i -= i & (-i))
                result += bit[i];
            return result;
        };
        
        int block = max(1, (int)sqrt(n));
        vector<int> order(q);
        iota(order.begin(), order.end(), 0);
        sort(order.begin(), order.end(), [&](int a, int b) {
            int ba = queries[a][0] / block, bb = queries[b][0] / block;
            if (ba != bb) return ba < bb;
            return (ba & 1) ? queries[a][1] > queries[b][1] : queries[a][1] < queries[b][1];
        });
        
        unordered_map<int, int> remCount;
        int numDistinctRem = 0;
        int curSize = 0;
        
        auto add = [&](int i) {
            if (remCount[rem[i]]++ == 0) numDistinctRem++;
            int idx = getIdx(quo[i]);
            updateBIT(countBIT, idx, 1);
            updateBIT(sumBIT, idx, quo[i]);
            curSize++;
        };
        
        auto removeFn = [&](int i) {
            if (--remCount[rem[i]] == 0) numDistinctRem--;
            int idx = getIdx(quo[i]);
            updateBIT(countBIT, idx, -1);
            updateBIT(sumBIT, idx, -quo[i]);
            curSize--;
        };
        
        auto getMedian = [&]() -> int {
            int target = (curSize + 1) / 2;
            int lo = 1, hi = m;
            while (lo < hi) {
                int mid = (lo + hi) / 2;
                if (queryBIT(countBIT, mid) >= target) hi = mid;
                else lo = mid + 1;
            }
            return sortedQuo[lo - 1];
        };
        
        auto computeOps = [&]() -> long long {
            if (numDistinctRem > 1) return -1;
            if (curSize <= 1) return 0;
            
            int median = getMedian();
            int medianIdx = getIdx(median);
            
            long long countLess = queryBIT(countBIT, medianIdx - 1);
            long long sumLess = queryBIT(sumBIT, medianIdx - 1);
            
            long long totalSum = queryBIT(sumBIT, m);
            long long countLessEq = queryBIT(countBIT, medianIdx);
            long long sumLessEq = queryBIT(sumBIT, medianIdx);
            
            long long countGreater = curSize - countLessEq;
            long long sumGreater = totalSum - sumLessEq;
            
            long long ops = (long long)median * countLess - sumLess + sumGreater - (long long)median * countGreater;
            
            return ops;
        };
        
        vector<long long> ans(q);
        int curL = 0, curR = -1;
        
        for (int i : order) {
            int l = queries[i][0], r = queries[i][1];
            
            while (curR < r) add(++curR);
            while (curL > l) add(--curL);
            while (curR > r) removeFn(curR--);
            while (curL < l) removeFn(curL++);
            
            ans[i] = computeOps();
        }
        
        return ans;
    }
};
# @lc code=end